fvm/src/call_manager/default.rs

```diff
@@ -433,30 +431,17 @@ where
             },
         };
 
-        // If a specific gas limit has been requested, create a child GasTracker and use that
-        // one hereon.
-        //
-        // The effectiveness of the user-supplied gas limit has been checked in the syscall handler.
-        // Here we are confident that the gas limit is below our gas available, and therefore
-        // we must enforce it.
-        let prev_gas_tracker = gas_limit.map(|gas_limit| {
-            let gas_used = self.gas_tracker.gas_used();
-            let tracing = self.machine.context().tracing;
-            let new = GasTracker::new(gas_limit, gas_used, tracing);
-            mem::replace(&mut self.gas_tracker, new)
-        });
+        // If a specific gas limit has been requested, try reduce the current limit (if the new
+        // limit is strictly less than the current limit).
+        let prev_gas_limit =
+            gas_limit.and_then(|gas_limit| self.gas_tracker.reduce_gas_remaining(gas_limit));
 
         // Do the actual send.
         let mut ret = self.send_resolved::<K>(from, to, method, params, value);
 
-        // Restore the original gas tracker and absorb the child's gas usage and traces into it.
-        if let Some(prev) = prev_gas_tracker {
-            let other = mem::replace(&mut self.gas_tracker, prev);
```

Comment:
Note: this didn't actually copy the gas _used_ (one of the reasons I prefer to keep the gas tracker and just edit the limits).

Reply:
Ah, nope. This was doing the right thing.